### PR TITLE
Add DescribeBackups Function

### PR DESCRIPTION
**Pull request by DeepikaDixit**
## Change Overview

Add `DescribeBackups` func which checks if `repo` is accessible, lists all `snapshots` in the `repo`, along with `repo` level `stats`

## Pull request type

Please check the type of change your PR introduces:
- [ ] Work in Progress
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Trivial/Minor
- [ ] Bugfix
- [X] Feature
- [ ] Documentation

## Test Plan

- [ ] Manual
- [x] Unit test
- [ ] E2E

### CODE REVIEW

**Comment by DeepikaDixit** [comment]
https://github.com/kastenhq/k10/pull/6670 is a draft implementation/use of `backupInfo` via K10

File: pkg/function/backups_info.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+	BackupsInfoRepoUnavailable   = "repoUnavailable"
+)
+
+func init() {
+	kanister.Register(&BackupsInfoFunc{})
+}
+
+var _ kanister.Func = (*BackupsInfoFunc)(nil)
+
+type BackupsInfoFunc struct{}
```
Comment (vkamra) [inline pkg/function/backups_info.go]: Not a fan of the function name. @DeepikaDixit @tdmanv - would `DescribeBackups` be better?

File: pkg/function/backups_info.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+				BackupsInfoRepoUnavailable:   "false",
+			},
+			nil
+	}
+}
+
+func (*BackupsInfoFunc) Exec(ctx context.Context, tp param.TemplateParams, args map[string]interface{}) (map[string]interface{}, error) {
+	var namespace, getBackupsInfoArtifactPrefix, encryptionKey string
+	var err error
+	if err = Arg(args, BackupsInfoNamespaceArg, &namespace); err != nil {
```
Comment (vkamra) [inline pkg/function/backups_info.go]: This should not be required

File: pkg/function/backups_info.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+var _ kanister.Func = (*BackupsInfoFunc)(nil)
+
+type BackupsInfoFunc struct{}
+
+func (*BackupsInfoFunc) Name() string {
+	return "BackupsInfo"
+}
+
+func BackupsInfo(ctx context.Context, cli kubernetes.Interface, tp param.TemplateParams, namespace, encryptionKey, targetPaths, jobPrefix string, podOverride sp.JSONMap) (map[string]interface{}, error) {
```
Comment (vkamra) [inline pkg/function/backups_info.go]: Do we need to export this method?

File: pkg/function/backups_info.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+		if err := kube.WaitForPodReady(ctx, cli, pod.Namespace, pod.Name); err != nil {
+			return nil, errors.Wrapf(err, "Failed while waiting for Pod %s to be ready", pod.Name)
+		}
+		pw, err := getPodWriter(cli, ctx, pod.Namespace, pod.Name, pod.Spec.Containers[0].Name, tp.Profile)
+		if err != nil {
+			return nil, err
+		}
+		defer cleanUpCredsFile(ctx, pw, pod.Namespace, pod.Name, pod.Spec.Containers[0].Name)
+		snapshotIDs, err := restic.GetSnapshotIDs(tp.Profile, cli, targetPath, encryptionKey, namespace, pod.Name, pod.Spec.Containers[0].Name)
+		if err != nil {
```
Comment (vkamra) [inline pkg/function/backups_info.go]: Minor - would a switch{} work better here instead?

File: pkg/function/backups_info.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+						BackupsInfoFileCount:         nil,
+						BackupsInfoSize:              nil,
+						BackupsInfoPasswordIncorrect: nil,
+						BackupsInfoRepoUnavailable:   "true",
+					},
+					nil
+			}
+			return nil, err
+		}
+		cmd, err := restic.StatsCommandByID(tp.Profile, targetPath, "" /* get all snapshot stats */, DefaultStatsMode, encryptionKey)
```
Comment (vkamra) [inline pkg/function/backups_info.go]: We're should get the repo size not the restore-size (`DefaultStatsMode` == `restore-size` )

File: pkg/function/backups_info.go (hunk 1 above)
Comment (tdmanv) [reply pkg/function/backups_info.go]: that works for me

File: pkg/function/backups_info.go (hunk 5 above)
Comment (DeepikaDixit) [reply pkg/function/backups_info.go]: 👍 Yep. Changing it to raw-data

File: pkg/function/backups_info.go (hunk 3 above)
Comment (DeepikaDixit) [reply pkg/function/backups_info.go]: You are right. I will change it

File: pkg/function/backups_info.go (hunk 1 above)
Comment (DeepikaDixit) [reply pkg/function/backups_info.go]: 👍

File: pkg/restic/restic_test.go [hunk 6]
```diff
@@ -286,5 +335,6 @@ func (s *ResticDataSuite) TestGetSnapshotStatsFromBackupLog(c *C) {
 		fc, s := SnapshotStatsFromBackupLog(tc.log)
 		c.Check(fc, Equals, tc.expectedfc)
 		c.Check(s, Equals, tc.expectedsize)
+
```
Comment (redgoat650) [inline pkg/restic/restic_test.go]: nit: get rid of newline?

File: pkg/function/backups_info.go (hunk 2 above)
Comment (DeepikaDixit) [reply pkg/function/backups_info.go]: Is this the correct call to get `controller ns`?
https://github.com/kanisterio/kanister/blob/6911c757d78870e04ff7120e6ca0bf28d8470290/pkg/kube/client.go#L36

Comment (vkamra) [reply pkg/function/backups_info.go]: No - I believe that returns the namespace set in the `current-context` in kube config.

`kube.GetControllerNamespace()` is the right method - see it's usage in `kube_task.go`

File: pkg/function/backups_info.go [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+const (
+	// DescribeBackupsNamespaceArg provides the namespace
+	DescribeBackupsNamespaceArg = "namespace"
+	// DescribeBackupsArtifactPrefixArg provides the path to restore backed up data
+	DescribeBackupsArtifactPrefixArg = "backupArtifactPrefix"
+	// DescribeBackupsEncryptionKeyArg provides the encryption key to be used for deletes
+	DescribeBackupsEncryptionKeyArg = "encryptionKey"
+	// DescribeBackupsPodOverrideArg contains pod specs to override default pod specs
+	DescribeBackupsPodOverrideArg    = "podOverride"
+	DescribeBackupsJobPrefix         = "get-backups-info-"
```
Comment (vkamra) [inline pkg/function/backups_info.go]: Minor - do not need to export this one

File: pkg/function/backups_info.go [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+			return nil, err
+		}
+		defer cleanUpCredsFile(ctx, pw, pod.Namespace, pod.Name, pod.Spec.Containers[0].Name)
+		snapshotIDs, err := restic.GetSnapshotIDs(tp.Profile, cli, targetPath, encryptionKey, namespace, pod.Name, pod.Spec.Containers[0].Name)
+		switch {
+		case err == nil:
+			break
+		case strings.Contains(err.Error(), restic.PasswordIncorrect):
+			return map[string]interface{}{
+					DescribeBackupsSnapshotIDs:       nil,
```
Comment (vkamra) [inline pkg/function/backups_info.go]: Should these keys be set with nil values or omitted completely?

File: pkg/function/backups_info.go [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+	// DescribeBackupsEncryptionKeyArg provides the encryption key to be used for deletes
+	DescribeBackupsEncryptionKeyArg = "encryptionKey"
+	// DescribeBackupsPodOverrideArg contains pod specs to override default pod specs
+	DescribeBackupsPodOverrideArg    = "podOverride"
+	DescribeBackupsJobPrefix         = "get-backups-info-"
+	DescribeBackupsFileCount         = "fileCount"
+	DescribeBackupsSize              = "size"
+	DescribeBackupsSnapshotIDs       = "snapshotIDs"
+	DescribeBackupsPasswordIncorrect = "passwordIncorrect"
+	DescribeBackupsRepoUnavailable   = "repoUnavailable"
```
Comment (vkamra) [inline pkg/function/backups_info.go]: Minor - rename to `DescribeBackupsRepoDoesNotExist`?

File: pkg/restic/restic_test.go [hunk 10]
```diff
@@ -238,6 +238,55 @@ func (s *ResticDataSuite) TestGetSnapshotStatsModeFromStatsLog(c *C) {
...
+func (s *ResticDataSuite) TestDoesRepoExist(c *C) {
+	for _, tc := range []struct {
+		log      string
+		expected bool
+	}{
+		{log: `Fatal: create key in repository at s3:s3.amazonaws.com/ddixit-test/testDir-dz4dv failed: repository master key and config already initialized`, expected: false},
+		{log: `Fatal: wrong password or no key found`, expected: false},
+		{log: `Fatal: unable to open config file: Stat: The specified key does not exist.
+Is there a repository at the following location?
+s3:s3.amazonaws.com/ddixit-test/testDir-dz`, expected: true},
```
Comment (vkamra) [inline pkg/restic/restic_test.go]: Let's not use real bucket names

File: pkg/function/backups_info.go (hunk 2 above)
Comment (DeepikaDixit) [reply pkg/function/backups_info.go]: 👍 Missed that. Thanks!

File: pkg/function/backups_info.go (hunk 8 above)
Comment (DeepikaDixit) [reply pkg/function/backups_info.go]: I think we can omit them.

**Comment by DeepikaDixit** [comment]
@vkamra Ready for merge?

File: pkg/function/data_test.go [hunk 11]
```diff
@@ -165,6 +166,27 @@ func newBackupDataBlueprint() *crv1alpha1.Blueprint {
...
+	return &crv1alpha1.Blueprint{
+		Actions: map[string]*crv1alpha1.BlueprintAction{
+			"describeBackups": &crv1alpha1.BlueprintAction{
+				Kind: param.StatefulSetKind,
+				Phases: []crv1alpha1.BlueprintPhase{
+					crv1alpha1.BlueprintPhase{
+						Name: "testDescribeBackups",
+						Func: "DescribeBackups",
+						Args: map[string]interface{}{
+							BackupDataNamespaceArg:            "{{ .StatefulSet.Namespace }}",
```
Comment (vkamra) [inline pkg/function/data_test.go]: Shouldn't need this

File: pkg/function/describe_backups.go [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+	kanister "github.com/kanisterio/kanister/pkg"
+	"github.com/kanisterio/kanister/pkg/format"
+	"github.com/kanisterio/kanister/pkg/kube"
+	"github.com/kanisterio/kanister/pkg/param"
+	"github.com/kanisterio/kanister/pkg/restic"
+)
+
+const (
+	// DescribeBackupsNamespaceArg provides the namespace
+	DescribeBackupsNamespaceArg = "namespace"
```
Comment (vkamra) [inline pkg/function/describe_backups.go]: Remove this

**Pull request merged**